web/src/pages/api/public/ingestion.ts

```diff
@@ -212,6 +212,36 @@ export default async function handler(
 
     const sortedBatch = sortBatch(batch);
 
+    // We group events by eventBodyId which allows us to store and process them
+    // as one which reduces infra interactions per event. Only used in the S3 case.
+    const sortedBatchByEventBodyId = sortedBatch.reduce(
+      (
+        acc: Record<
+          string,
+          {
+            data: IngestionEventType[];
+            key: string;
+            type: (typeof eventTypes)[keyof typeof eventTypes];
+          }
+        >,
+        event,
+      ) => {
+        if (!event.body?.id) {
+          return acc;
+        }
+        if (!acc[event.body.id]) {
+          acc[event.body.id] = {
+            data: [],
+            key: event.id,
+            type: event.type,
+          };
+        }
+        acc[event.body.id].data.push(event);
```

Comment:
We could also have the same id for different types (e.g. scores, traces, observations). This would be valid. We need to ensure that they are processed separately. Hence checking for type + id is probably better here.

Reply:
@maxdeichmann Covered in https://github.com/langfuse/langfuse/pull/3770/commits/17ee7ed2aacfca88880093a63c66e9e6793fdadd.